clang-tools-extra/clangd/XRefs.cpp

```diff
@@ -2292,9 +2289,26 @@ incomingCalls(const CallHierarchyItem &Item, const SymbolIndex *Index) {
   Index->lookup(ContainerLookup, [&](const Symbol &Caller) {
     auto It = CallsIn.find(Caller.ID);
     assert(It != CallsIn.end());
-    if (auto CHI = symbolToCallHierarchyItem(Caller, Item.uri.file()))
+    if (auto CHI = symbolToCallHierarchyItem(Caller, Item.uri.file())) {
+      SymbolLocation CallerLoc =
+          Caller.Definition ? Caller.Definition : Caller.CanonicalDeclaration;
+      std::vector<Range> FromRanges;
+      for (const SymbolLocation &L : It->second) {
+        if (StringRef{L.FileURI} != StringRef{CallerLoc.FileURI}) {
+          elog("incomingCalls: call location not in same file as caller, this "
```

Comment:
Should we assert as well?

Reply:
I think this case is possible.

```
// t.h
#pragma  once
#define abc2 void test4() {

// t.cc
#include "./t.h"

void func();

abc 
  func();
}
```

calling call-hierarchy for `func()`.

Reply:
Thanks for the example! This is exactly the sort of case the hardening was intended to catch, though I didn't think of it.

The case raises an interesting question: what **should** our behaviour be for an example like this, where the function definition is spread across multiple files?

The [protocol representation](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#callHierarchy_incomingCalls) only allows specifying one file per caller:

```ts
export interface CallHierarchyIncomingCall {

	/**
	 * The item that makes the call.
	 */
	from: CallHierarchyItem;

	/**
	 * The ranges at which the calls appear. This is relative to the caller
	 * denoted by [`this.from`](#CallHierarchyIncomingCall.from).
	 */
	fromRanges: Range[];
}

```

Here, `from.uri` is the only thing that encodes a file; the ranges in `fromRanges` encode a pair of line/column positions only, and are assumed to be in the same file as `from.uri`.

Reply:
I don't have a good solution, it looks like the LSP protocol assumes that the ranges are in the same file as from.uri. I think it is worth to reporting it to LSP.

It is not a common case though, I think we could just drop these ranges in the current implementation.